### DISCUSSION TITLE
[2024-10-28] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by vitvakatu** [enso-org/enso#11425] [Weekly Demo Day]
- https://github.com/enso-org/enso/pull/11398
- https://github.com/enso-org/enso/pull/11421
- Reviewing and helping with the testing of https://github.com/enso-org/enso/pull/11337
- https://github.com/enso-org/enso/issues/11222

https://github.com/user-attachments/assets/a4d711be-5526-41aa-8e24-6be52e187844

**Comment by farmaazon** [comment]
* Merged #11271
* #11332

[Screencast From 2024-10-17 13-51-45.webm](https://github.com/user-attachments/assets/78924f87-6bdb-4cb6-8f95-d9f97c63aea4)

* #11358

[Screencast From 2024-10-18 15-23-40.webm](https://github.com/user-attachments/assets/cdc3e5f0-05f3-4858-ac98-c634ca283244)

* #11388

![image](https://github.com/user-attachments/assets/995d1a85-eb04-4b8d-bf23-b47ea61185e4)

* Also a bit of maintenance work, bumped eslint and fixed flaky tests in  #11386
* In progress: storing widget size in source code metadata.

**Comment by Frizi** [comment]
## 🍃 BAZEL progress

- Bazel build workflow is now successfuly running on github actions using github's linux runners. The build produces a valid downloadable artefact containing all GUI static assets.
    https://github.com/enso-org/enso/actions/workflows/bazel-build.yml
- Split the gui environment configuration away from the main build artifacts. The same `vite` command output can be easily transformed into testing or production package without rebuilding. This is done in preparation for "promote to production" workflow, guaranteeing that exactly the same code is used across testing builds and actual releases.
- Completely removed the dependency on locally installed rust or wasm toolchains (e.g. wasm-bindgen), even when building the GUI locally. All rust-related artifacts are built using bazel **even in local development**.

Remaining work:
- Some dashboard tests currently fail, and that need to be addressed.
- The bazel build currently fails on Windows, which also affects our CI runners. This need to be sorted before merging is possible.
- Merging this PR proves to be problematic due to frequent merge conflicts, which in turn introduce linting and testing errors.

**Comment by kazcw** [comment]
Parser improvements: "What is a function, really?"
```ruby
## ICON data_input
   Returns a new Connection with the specified database set as default.
@database (Single_Choice display=Display.Always values=[Option 'Nothing'])
set_database : Text -> Connection ! SQL_Error
set_database self database =
    if database == self.database then self else SQL_Error.throw_sql_error "Changing database is not supported."
```

*In the new AST representation, this is all one `Function` node, with fields for all the types of line that contribute to a function definition.*

Moving the assembly of full function definitions into the parser:
- Allows simplifying the GUI and backend logic (fixing some current GUI bugs)
- Allows the parser to catch errors when modifiers aren't able to be attached to an item

Merged PRs (4/5):
- #11324 
  - Fixes GUI handling of 0-argument functions
- #11346 
  - Also introduce forward-compatible API for parser test cases
- #11364 
  - #11293
    - Fixes GUI handling of block with type-annotated tail.
  - Fixes GUI inserting a function "before" a function that has a type-sig line (i.e. when collapsing components).
- #11374 
  - Also enables multiline constructor definitions

Currently finishing last step:
- #11302

Misc:
- #11139

**Comment by hubertp** [comment]
Relevant backend changes:
- Metadata should not depend on absolute text spans #11390

To discuss:
Two file edits on startup, which seem unnecessary.
